soc/arm/nxp_kinetis/k2x/soc.c

```diff
@@ -0,0 +1,210 @@
+		.enableMode = 0U, /* Disable external reference clock */
+		.erclkDiv = 0U,
+	},
+};
+
+static const mcg_pll_config_t pll0Config = {
+	.enableMode = 0U,
+	.prdiv = CONFIG_MCG_PRDIV0,
+	.vdiv = CONFIG_MCG_VDIV0,
+};
+
+static const sim_clock_config_t simConfig = {
+	.pllFllSel = PLLFLLSEL_MCGPLLCLK, /* PLLFLLSEL select PLL. */
+	.er32kSrc = ER32KSEL_RTC,         /* ERCLK32K selection, use RTC. */
+	.clkdiv1 = SIM_CLKDIV1_OUTDIV1(CONFIG_K22_CORE_CLOCK_DIVIDER - 1) |
+		   SIM_CLKDIV1_OUTDIV2(CONFIG_K22_BUS_CLOCK_DIVIDER - 1) |
+		   SIM_CLKDIV1_OUTDIV3(CONFIG_K22_FLEXBUS_CLOCK_DIVIDER - 1) |
+		   SIM_CLKDIV1_OUTDIV4(CONFIG_K22_FLASH_CLOCK_DIVIDER - 1),
+};
+
+/**
+ *
+ * @brief Initialize the system clock
+ *
+ * This routine will configure the multipurpose clock generator (MCG) to
+ * set up the system clock.
+ * The MCG has nine possible modes, including Stop mode.  This routine assumes
+ * that the current MCG mode is FLL Engaged Internal (FEI), as from reset.
+ * It transitions through the FLL Bypassed External (FBE) and
+ * PLL Bypassed External (PBE) modes to get to the desired
+ * PLL Engaged External (PEE) mode and generate the maximum 120 MHz system
+ * clock.
+ *
+ * @return N/A
+ *
+ */
+static ALWAYS_INLINE void clkInit(void)
+{
+/* Set HSRUN power mode */
+	SMC_SetPowerModeProtection(SMC, kSMC_AllowPowerModeAll);
+	SMC_SetPowerModeHsrun(SMC);
+	while (SMC_GetPowerModeState(SMC) != kSMC_PowerStateHsrun){
+	}
+
+	CLOCK_SetSimSafeDivs();
+
+	CLOCK_InitOsc0(&oscConfig);
+	CLOCK_SetXtal0Freq(CONFIG_OSC_XTAL0_FREQ);
+
+
+	CLOCK_SetInternalRefClkConfig(kMCG_IrclkEnable, kMCG_IrcSlow,
+				      CONFIG_MCG_FCRDIV);
+
+	/* Configure FLL external reference divider (FRDIV). */
+	/*CLOCK_CONFIG_SetFllExtRefDiv(0);*/
+	MCG->C1 = ((MCG->C1 & ~MCG_C1_FRDIV_MASK) | MCG_C1_FRDIV(0x0U));
+
+	CLOCK_BootToPeeMode(kMCG_OscselOsc, kMCG_PllClkSelPll0, &pll0Config);
+
+	CLOCK_SetSimConfig(&simConfig);
+
+#if CONFIG_USB_KINETIS
+	CLOCK_EnableUsbfs0Clock(kCLOCK_UsbSrcPll0,
+				CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+#endif
+}
+
+/**
+ *
+ * @brief Perform basic hardware initialization
+ *
+ * Initialize the interrupt controller device drivers.
+ * Also initialize the timer device driver, if required.
+ *
+ * @return 0
+ */
+
+static int fsl_frdm_k22f_init(struct device *arg)
+{
+	ARG_UNUSED(arg);
+
+	unsigned int oldLevel; /* old interrupt lock level */
+#if !defined(CONFIG_ARM_MPU)
+#if defined(SYSMPU)
+	u32_t temp_reg;
+#endif
+#endif /* !CONFIG_ARM_MPU */
+
+	/* disable interrupts */
+	oldLevel = irq_lock();
+
+	/* release I/O power hold to allow normal run state */
+	PMC->REGSC |= PMC_REGSC_ACKISO_MASK;
+
+#if !defined(CONFIG_ARM_MPU)
+	/*
+	 * Disable memory protection and clear slave port errors.
+	 * Note that the K22F does not implement the optional ARMv7-M memory
+	 * protection unit (MPU), specified by the architecture (PMSAv7), in the
+	 * Cortex-M4 core.  Instead, the processor includes its own MPU module.
+	 */
+#if defined(SYSMPU)
```

Comment:
Why is this extra condition here compared to `soc/arm/nxp_kinetis/k6x/soc.c`?

Reply:
There is no SYSMPU register in the HAL layer for the k22, but since it's listed as an option on nxp.com so I have kept the functionality